book/src/leader-rotation.md

```diff
@@ -3,17 +3,92 @@
+of the slots it assigns, such that the ledger state it uses to compute the
+schedule is finalized. That duration is called the *leader schedule offset*.
+Solana sets the offset to the duration of slots until the next epoch. That is,
+the leader schedule for an epoch is calculated from the ledger state at the
+start of the previous epoch. The offset of one epoch is fairly arbitrary and
+assumed to be sufficiently long such that all validators will have finalized
+their ledger state before the next schedule is generated. A cluster may choose
+to shorten the offset to reduce the time between stake changes and leader
+schedule updates.
+
+While operating without partitions lasting longer than an epoch the schedule
+only needs to be generated when the root fork crosses the epoch boundary.  Since
+the schedule is for the next epoch, any new stakes committed to the root fork
+will not be active until the next epoch.
+
+Without a partition lasting longer then an epoch, the cluster will work as
+follows:
+
+1. A validator continuously updates its own root fork as it votes.
+
+2. The validator updates its leader schedule each time the slot height crosses
+an epoch boundary.
+
+For example:
+
+The epoch duration is 100 slots. The root fork is updated from fork computed at
+slot height 99 to a fork computed at slot height 102. Forks with slots at height
+100,101 were skipped because of failures.  The new leader schedule is computed
+using fork at slot height 102.  It is active from slot 200 until it is updated
+again.
+
+No inconsistency can exist because every validator that is voting with the
+cluster has skipped 100 and 101 when its root reaches 102.  The duration of the
+*leader schedule offset* has a direct relationship to the likelihood of a
+cluster inconsistency state.  If the *leader schedule offset* is 6 standard
+deviations longer than the median partition time, then the likelihood of a
+partition longer than the offset is 1 in 1 million.
+
+### Leader Schedule Rotation with Epoch Sized Partitions.
+
+Consider the following scenario:
+
+Two partitions that are generating half of the blocks each.  Neither is coming
+to a definitive supermajority fork.  Both will cross epoch 100 and 200 without
+actually committing to a root and therefore a cluster wide commitment to a new
+leader schedule.
+
+In this unstable scenario, multiple valid leader schedules exist.
+
+* A leader schedule is generated for every fork whose direct parent is in the
+previous epoch.
+
+* The leader schedule is valid after the start of the next epoch for descendant
+forks until it is updated.
+
+Each partition's schedule will diverge after the partition lasts more than an
+epoch.  For this reason, the epoch duration should be selected to be much much
+larger then slot time and the expected length for a fork to be committed to
+root.
+
+## Leader Schedule Generation at Genesis
+
+The genesis block declares the first leader.  This leader is scheduled for the
```

Comment:
It'd be a bit more elegant to have the system generate the second epoch's leader schedule as part of the normal algorithm. If the offset is one epoch, it should fire immediately on startup (slot height 0).

Reply:
@garious wouldn't that result in the same schedule, since at slot 0 it would still be the first leader?

Reply:
Yes, that's the point. That's why it's more elegant. At slot 0, there is no need for special cases. You'll get the same result as if you postponed generating the first active set until epoch height 1.